packages/contracts/contracts/PriceAdapter.sol

```diff
@@ -0,0 +1,207 @@
+
+  constructor(address _baseDataFeed, address _quoteDataFeed) {
+    baseDataFeed = AggregatorV3Interface(_baseDataFeed);
+    quoteDataFeed = AggregatorV3Interface(_quoteDataFeed);
+  }
+
+  function factory() external view override returns (address) {}
+
+  function token0() external view override returns (address) {}
+
+  function token1() external view override returns (address) {}
+
+  function fee() external view override returns (uint24) {}
+
+  function tickSpacing() external view override returns (int24) {}
+
+  function maxLiquidityPerTick() external view override returns (uint128) {}
+
+  function slot0()
+    external
+    view
+    override
+    returns (
+      uint160 sqrtPriceX96,
+      int24 tick,
+      uint16 observationIndex,
+      uint16 observationCardinality,
+      uint16 observationCardinalityNext,
+      uint8 feeProtocol,
+      bool unlocked
+    )
+  {}
+
+  function feeGrowthGlobal0X128() external view override returns (uint256) {}
+
+  function feeGrowthGlobal1X128() external view override returns (uint256) {}
+
+  function protocolFees() external view override returns (uint128 _token0, uint128 _token1) {}
+
+  function liquidity() external view override returns (uint128) {}
+
+  function ticks(
+    int24 tick
+  )
+    external
+    view
+    override
+    returns (
+      uint128 liquidityGross,
+      int128 liquidityNet,
+      uint256 feeGrowthOutside0X128,
+      uint256 feeGrowthOutside1X128,
+      int56 tickCumulativeOutside,
+      uint160 secondsPerLiquidityOutsideX128,
+      uint32 secondsOutside,
+      bool initialized
+    )
+  {}
+
+  function tickBitmap(int16 wordPosition) external view override returns (uint256) {}
+
+  function positions(
+    bytes32 key
+  )
+    external
+    view
+    override
+    returns (
+      uint128 _liquidity,
+      uint256 feeGrowthInside0LastX128,
+      uint256 feeGrowthInside1LastX128,
+      uint128 tokensOwed0,
+      uint128 tokensOwed1
+    )
+  {}
+
+  function observations(
+    uint256 index
+  )
+    external
+    view
+    override
+    returns (uint32 blockTimestamp, int56 tickCumulative, uint160 secondsPerLiquidityCumulativeX128, bool initialized)
+  {}
+
+  function observe(
+    uint32[] calldata secondsAgos
+  )
+    external
+    view
+    override
+    returns (int56[] memory tickCumulatives, uint160[] memory secondsPerLiquidityCumulativeX128s)
+  {
+    secondsPerLiquidityCumulativeX128s = new uint160[](secondsAgos.length);
+    secondsPerLiquidityCumulativeX128s[0] = secondsAgos[0];
+    secondsPerLiquidityCumulativeX128s[1] = secondsAgos[1];
+
+    uint160 sqrtPriceX96 = getSqrtPriceX96();
+    tickCumulatives = OracleLib.getCumulativeTickAtSqrtRatio(sqrtPriceX96, secondsAgos);
+  }
+
+  function getSqrtPriceX96() public view returns (uint160) {
+    (int256 price, int256 decimals) = getPrice();
+
+    uint160 sqrtPriceX96 = uint160((OracleLib.sqrt(((uint256(price) << 96) / uint256(decimals))) << 48));
```

Comment:
1. 96 can be changed to FP96.RESOLUTION
2. The order of operations `sqrt, division` should be better precision-wise than current `division, sqrt`.